src/common/history.c

```diff
@@ -760,8 +765,9 @@ int dt_history_copy_and_paste_on_image(int32_t imgid, int32_t dest_imgid, gboole
     return 1;
   }
 
-  // Just in case lock the database
-  dt_pthread_mutex_lock(&darktable.db_insert);
+  dt_pthread_mutex_lock(&darktable.db_multi_image);
+  dt_database_lock_image(imgid);
```

Comment:
This one IS dangerous. Can't we have a scenario where we have the opposite lock scenario?

```
   dt_database_lock_image(des_imgid);
   dt_database_lock_image(imgid);
```

Reply:
I've already covered this issue. @jenshannoschwalm tries to solves this problem by adding one more mutex - db_multi_image. It seems to solve the deadlock between imgid and dest_imgid, but it may introduce other deadlocks down the call stack, I'm not sure here. My proposed workaround didn't add new mutexes. I really don't think it's a good idea to solve the problems of multiple mutex deadlock by adding more mutexes;)

Reply:
Also my proposed solution doesn't introduce another global image lock after we just removed one;)

Reply:
I think your suggestion does not work. Binding two mutex locks/unlocks in a function one after another does not help as jobA(1,3) and jobB(3,1) would possible deadlock. Either check visually by code analysis :-) or prepend a global mutex in **all** functions that will lock two images. That's my idea.

Reply:
Rethinking here, i still think your proposal is not ok but introducing a global lock is just bad. Need to find another way. What about this?

We have a pool of known image id's that are not used by dt itself. (lets say last 64imgs in the uint32 range.) By trylocking we can find a not used id.

If we have a multi-image function using img1 and img2 we lock img1, get a still unlocked and right now locked imgx from pool, copy database fom img1->imgx, unlock img1, lock img2, continue by using imgx instead of img1, and at the end of the function unlock img2 and imgx.

Reply:
Please stop complicating the logic. You won't solve the issue by making it harder to understand.
My proposal works. Please re-read it and think of counter-example if you still think it doesn't work.